pulser-core/pulser/math/abstract_array.py

```diff
@@ -236,6 +241,10 @@ def __setitem__(self, indices: Any, values: AbstractArrayLike) -> None:
         self._array = array
         del self.is_tensor  # Clears cache
 
+    def __iter__(self) -> Generator:
+        for i in self._array:
+            yield i
+
```

Comment:
I think that a class with `__getitem__` and `__len__` is automatically iterable. What's the usecase for defining `__iter__` explicitly?

Reply:
I got some `mypy` errors related to `AbstractClass` not having `__iter__` . For example, when passing this object to `zip`: `for name, (x, y) in zip(names, self._coords)`.

Reply:
Ok... Well, if we are going to define `__iter__` we have to do it properly. I propose something like this: 
```suggestion
    def __iter__(self) -> Generator[AbstractArray, None, None]]:
        for i in range(self.__len__()):
            yield self.__getitem__(i)

```

Reply:
If I include the specification `[AbstractArray, None, None]` for `Generator` type I get these `mypy` errors:
```
pulser-core/pulser/register/base_register.py:91: error: Argument 2 to "zip" has incompatible type "Union[AbstractArray, List[Any]]"; expected "Iterable[ndarray[Any, Any]]"  [arg-type]
pulser-core/pulser/register/base_register.py:261: error: Argument 1 to "tolist" of "_ArrayOrScalarCommon" has incompatible type "Union[AbstractArray, Any]"; expected "_ArrayOrScalarCommon"  [arg-type]
```

Reply:
These errors look sensible to me, the type hints in `BaseRegister` need to be updated or the methods changed

Reply:
For example, I would change the `qubits` property to:

```python
    @property
    def qubits(self) -> dict[QubitId, pm.AbstractArray]:
        """Dictionary of the qubit names and their position coordinates."""
        return dict(zip(self._ids, pm.vstack(self._coords)))
```